src/cmake/build_ZLIB.cmake

```
@@ -3,36 +3,131 @@
+# Set ENABLE_ZLIBNG=OFF to build the original zlib.
+check_is_enabled(ZLIBNG ENABLE_ZLIBNG)
+
+
+# Set the version to build; note that 1.x versions will build the original zlib,
+# while 2.x or later will build zlib-ng.
+if (ENABLE_ZLIBNG)
+    set_cache (ZLIB_BUILD_VERSION ${ZLIBNG_VERSION_LATEST} "ZLIB or ZLIB-NG version for local builds")
+else ()
+    set_cache (ZLIB_BUILD_VERSION ${ZLIB_VERSION_LATEST} "ZLIB or ZLIB-NG version for local builds")
+endif ()
+
+
+# Choose the git repository, tag format, and extra arguments based on version.
+# For now, we're assuming that the original zlib will never reach version 2.0, 
+# so anything greater than or equal to 2.0 is zlib-ng.
+if (ZLIB_BUILD_VERSION VERSION_LESS "2.0.0")
+  # Original zlib: use the 'v' prefix in the tag.
+  set (ZLIB_GIT_REPOSITORY "https://github.com/madler/zlib")
+  set (ZLIB_GIT_TAG "v${ZLIB_BUILD_VERSION}")
+  set (ZLIB_BUILD_OPTIONS "")
+  set (ZLIBNG_USED FALSE)
+  if (ENABLE_ZLIBNG)
+    message (STATUS "Building zlib version ${ZLIB_BUILD_VERSION}, even though ENABLE_ZLIBNG=${ENABLE_ZLIBNG}")
+    message (STATUS "If you believe this is a mistake, check usages of `checked_find_package(ZLIB ...)` in the code base.")
+    message (STATUS "See src/cmake/build_ZLIB.cmake for more details.")
+  else ()
+    message (STATUS "Building zlib version ${ZLIB_BUILD_VERSION}")
+  endif ()
+else ()
+  # zlib-ng: omit the 'v' prefix for the git tag and enable compatibility mode.
+  set (ZLIB_GIT_REPOSITORY "https://github.com/zlib-ng/zlib-ng")
+  set (ZLIB_GIT_TAG "${ZLIB_BUILD_VERSION}")
+  set (ZLIB_BUILD_OPTIONS "-DZLIB_COMPAT=ON;-DWITH_GTEST=OFF;-DWITH_GZFILEOP=OFF;-DZLIBNG_ENABLE_TESTS=OFF;-DZLIB_ENABLE_TESTS=OFF")
+  set (ZLIBNG_USED TRUE)
+  message (STATUS "Building zlib-ng version ${ZLIB_BUILD_VERSION}")
+endif ()
+
+
+set_cache (ZLIB_BUILD_SHARED_LIBS OFF
+  DOC "Should execute a local ZLIB build, if necessary, build shared libraries" ADVANCED)
```

Concern: by building zlib-ng as a static library, won't that prevent it from being used by libpng, which is where much of the benefit of the faster zlib implementation would come from? Any static libraries we incorporate into libOpenImageIO have their symbols hidden by default, specifically so nothing else we link will find those symbols.

---

Oh yeah? You'd definitely know better than I; I only barely understand what I'm doing! But yes, we specifically want PNG to find and use zlib-ng; the rest doesn't really matter.

I was hoping by placing the local build prefix at the front of CMAKE_PREFIX_PATH, and passing CMAKE_IGNORE_PATH to child build processes, that would coerce the PNG build process to find and link zlib-ng before any other zlibs. I'll should turn on shared libs and turn off plugin bundling to see what's actually happening...

---

Hmmm, I see what you mean. I'm honestly not sure. You might be right that it could be picked up by libpng and other dependencies IF we build them ourselves. But I think the usual case is that some/most/all of the dependencies are already built.

One possibility is that we *should* build statically, but expose the symbols by adding them to the global list in hidesymbols.map, so that libpng and others will find them inside our library at actual load time. We may need to also modify libOpenImageIO/CMakeLists where we have `-Wl,--exclude-libs,ALL`. (By default, we hide all symbols from static library dependencies so that we only use them internally but don't expose their symbols for dependencies to link against.)

But once we expose them, they might be picked up by anything that links us as a dependency, too, depending on the link order. So, for example, an app that itself uses zlib and OIIO might end up itself using zlib-ng inadvertently. Is that always ok? Maybe it's fine, I just think we need to be confident that this is the behavior we want.

---

> So, for example, an app that itself uses zlib and OIIO might end up itself using zlib-ng inadvertently. Is that always ok? Maybe it's fine, I just think we need to be confident that this is the behavior we want.

Hmm. Very interesting. If the zlib-ng documentation is to be believed, I would imagine we'd want this behavior; but I don't feel qualified to so say so for sure...

Would exposing zlib-ng-compatibility-mode symbols in OIIO impose unexpected constraints or build-time requirements on building apps linking libraries that themselves link OIIO?

---

Update: @joaovbs96 tested one of the [Windows bdist artifacts](https://github.com/AcademySoftwareFoundation/OpenImageIO/actions/runs/13337407397) produced by the last Wheels run for this thread, and has confirmed PNG write speed improvements consistent with having linked against zlib-ng instead of zlib -- in his use case (which inspired this PR), write times for very large PNGs decreased from ~20s to ~8s.